File: 016.NNCoreAnimation/NNCoreAnimation/Controllers/NNPositionViewController.m
//
//  NNPositionViewController.m
//  NNCoreAnimation
//
//  Created by 刘朋坤 on 2018/8/13.
//  Copyright © 2018年 刘朋坤. All rights reserved.
//

#import "NNPositionViewController.h"

@interface NNPositionViewController ()<CAAnimationDelegate>

@end

@implementation NNPositionViewController

- (NSArray *)buttonTitleArray {
    return @[@"top", @"bottom", @"left", @"right", @"keyFrame", @"path"];
}

- (void)buttonClick:(UIButton *)sender {
    switch (sender.tag) {
        case 0:
            [self positionAnimationWithToValue:CGPointMake(self.nnView.headerImage.center.x, self.nnView.headerImage.center.y - 100)];
            break;
        case 1:
            [self positionAnimationWithToValue:CGPointMake(self.nnView.headerImage.center.x, self.nnView.headerImage.center.y + 100)];
            break;
        case 2:
            [self positionAnimationWithToValue:CGPointMake(self.nnView.headerImage.center.x-100, self.nnView.headerImage.center.y)];
            break;
        case 3:
            [self positionAnimationWithToValue:CGPointMake(self.nnView.headerImage.center.x+100, self.nnView.headerImage.center.y)];
            break;
        case 4:
            [self keyFrameAnimation];
            break;
        case 5:
            [self pathAnimation];
            break;
            
        default:
            break;
    }
}

- (void)positionAnimationWithToValue:(CGPoint)toValue {
    CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:@"position"];
    animation.fromValue = [NSValue valueWithCGPoint:self.nnView.headerImage.center];
    animation.toValue = [NSValue valueWithCGPoint:toValue];
    animation.duration = 1.5;
    animation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseInEaseOut];
    [self.nnView.headerImage.layer addAnimation:animation forKey:@"positionAnimation"];
}

- (void)keyFrameAnimation {
    CAKeyframeAnimation *animation = [CAKeyframeAnimation animationWithKeyPath:@"position"];
    NSValue *value1 = [NSValue valueWithCGPoint:CGPointMake(100, 100)];
    NSValue *value2 = [NSValue valueWithCGPoint:CGPointMake(200, 300)];
    NSValue *value3 = [NSValue valueWithCGPoint:CGPointMake(500, 100)];
    NSValue *value4 = [NSValue valueWithCGPoint:CGPointMake(100, 500)];
    NSValue *value5 = [NSValue valueWithCGPoint:CGPointMake(300, 100)];
    animation.values = [NSArray arrayWithObjects:value1, value2, value3, value4, value2, value5, nil];
    animation.duration = 3.0f;
    animation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseInEaseOut];
    animation.delegate = self;
    [self.nnView.headerImage.layer addAnimation:animation forKey:@"keyFrameAnimation"];
}

- (void)pathAnimation {
    CAKeyframeAnimation *animation = [CAKeyframeAnimation animationWithKeyPath:@"position"];
    UIBezierPath *path = [UIBezierPath bezierPathWithOvalInRect:CGRectMake(self.nnView.headerImage.center.x-75, self.nnView.headerImage.center.y-75, 150, 150)];
    animation.path = path.CGPath;
    animation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseInEaseOut];
    animation.repeatCount = 2;
    animation.duration = 2.0f;
    [self.nnView.headerImage.layer addAnimation:animation forKey:@"pathAnimation"];
}

@end

File: 018.NNCoreAnimation_swift/NNCoreAnimation_swift/NNCoreAnimation_swift/Controllers/NNPositionViewController.swift
//
//  NNPositionViewController.swift
//  NNCoreAnimation_swift
//
//  Created by 刘朋坤 on 2019/1/5.
//  Copyright © 2019 刘朋坤. All rights reserved.
//

import UIKit

class NNPositionViewController: NNBaseViewController {

    // 按钮点击事件
    override func buttonClick(index: NSInteger) {
        switch index {
        case 0:
            positionAnimationWithToValue(toValue: CGPoint.init(x: self.nnView.headerImage.center.x, y: self.nnView.headerImage.center.y - 100))
            break
        case 1:
            positionAnimationWithToValue(toValue: CGPoint.init(x: self.nnView.headerImage.center.x, y: self.nnView.headerImage.center.y + 100))
            break
        case 2:
            positionAnimationWithToValue(toValue: CGPoint.init(x: self.nnView.headerImage.center.x-100, y: self.nnView.headerImage.center.y))
            break
        case 3:
            positionAnimationWithToValue(toValue: CGPoint.init(x: self.nnView.headerImage.center.x+100, y: self.nnView.headerImage.center.y))
            break
        case 4:
            keyFrameAnimation()
            break
        case 5:
            pathAnimation()
            break
        default:
            break
        }
    }
    
    func positionAnimationWithToValue(toValue: CGPoint) {
        let animation = CABasicAnimation.init(keyPath: "position")
        animation.fromValue = NSValue.init(cgPoint: self.nnView.headerImage.center)
        animation.toValue = NSValue.init(cgPoint: toValue)
        animation.duration = 1.5
        animation.timingFunction = CAMediaTimingFunction.init(name: .easeInEaseOut)
        self.nnView.headerImage.layer.add(animation, forKey: "positionAnimation")
    }
    
    func keyFrameAnimation() {
        let animation = CAKeyframeAnimation.init(keyPath: "position")
        let value1 = NSValue.init(cgPoint: CGPoint.init(x: 100, y: 100))
        let value2 = NSValue.init(cgPoint: CGPoint.init(x: 200, y: 300))
        let value3 = NSValue.init(cgPoint: CGPoint.init(x: 500, y: 100))
        let value4 = NSValue.init(cgPoint: CGPoint.init(x: 100, y: 500))
        let value5 = NSValue.init(cgPoint: CGPoint.init(x: 300, y: 100))
        animation.values = [value1, value2, value3, value4, value5]
        animation.duration = 3
        animation.timingFunction = CAMediaTimingFunction.init(name: .easeInEaseOut)
        self.nnView.headerImage.layer.add(animation, forKey: "keyFrameAnimation")
    }
    
    func pathAnimation() {
        let animation = CAKeyframeAnimation.init(keyPath: "position")
        let path = UIBezierPath.init(rect: CGRect.init(x: self.nnView.headerImage.center.x-75, y: self.nnView.headerImage.center.y-75, width: 150, height: 150))
        animation.path = path.cgPath
        animation.timingFunction = CAMediaTimingFunction.init(name: .easeInEaseOut)
        animation.repeatCount = 2
        animation.duration = 2
        self.nnView.headerImage.layer.add(animation, forKey: "pathAnimation")
    }
    
    // 数组元素
    override func buttonTitleArrayMethod() -> NSArray {
        return ["top", "bottom", "left", "right", "keyFrame", "path"]
    }

}
